### DISCUSSION TITLE
How context_menu does work?

### DISCUSSION THREAD

**Discussion by Zeroto521** [Flow-Launcher/Flow.Launcher#574] [Q&A] [answered]
https://github.com/Zeroto521/Wox.Plugin.Bark/

> It is better to paste the message in Wox. If you input the message one by one, Wox would send the message one by one also. There are not Enter in Wox. ¯\_(ツ)_/¯

In such an application, it needs a **button** to send.

The following one is what I want, for a folder we could enter the context_menu and choose to open it at explorer.

![image](https://user-images.githubusercontent.com/25895405/126070224-37b8e455-ac09-4300-87ea-44d78653b5c4.png)
![image](https://user-images.githubusercontent.com/25895405/126070230-6ee30b77-5d35-414f-b523-22389a7cac11.png)

---

From the source, we could see `context_menu` like to `query` need an argument.

https://github.com/Flow-Launcher/Flow.Launcher.JsonRPC.Python/blob/1e95f0df1853bee026c3bfea64ff48ea133db884/flowlauncher/FlowLauncher.py#L35-L45

```python
    def query(self, param: str = '') -> list:
        """
        sub class need to override this method
        """
        return []


    def context_menu(self, data) -> list:
        """
        optional context menu entries for a result
        """
        return []
```

For `query`, the argument comes from our input.
Like this, the argument is `flow` from the querying box.
![image](https://user-images.githubusercontent.com/25895405/126070224-37b8e455-ac09-4300-87ea-44d78653b5c4.png)

But for `context_menu`, the argument from where?
To finish opening the folder at the explorer, I thought there should pass some information from the `query` result to this result's context_menu.
Such as the path of the folder.

---

PS:

Wox seems nothing happened when I tested it.

Flow v1.7.2 has a bug about context_menu (deefrawley/Flow.Launcher.Plugin.Currency#9). 
So I'm hard to test this idea.

**Comment by taooceros** [comment]
Let me take a look on the code. I remember the argument is inside the return json created from python plugin.

**Comment by pc223** [comment, marked as answer]
Not sure if this is what you looking for, you can try add "ContextData" in **response of query()**, after that, when you enter context menu mode, the `data` variable (`def context_menu(self, data) -> list:`) will be `[["param1","param2","param3"]]`


```python
{
    "result":[
        {
            "Title":"title1"
            "SubTitle":"subtitle1",
            "IcoPath":"assets/favicon.ico",
            "ContextData":["param1","param2","param3"],
        }
    ]
}
```

**Reply by pc223** [reply]
@taooceros Oh! same time! 😝

**Reply by pc223** [reply]
But I'm not too sure if this is intended usage, @taooceros can you confirm if this is the "intended" way to use "ContextData"? I don't see this in the docs 😅

**Reply by taooceros** [reply]
> But I'm not too sure if this is intended usage, @taooceros can you confirm if this is the "intended" way to use "ContextData"? I don't see this in the docs 😅 

To store information for the specific result. Because context menu is for one specific results, so to display the context menu of the result, we need some information from the result.

In some Flow built in plugin, the context data is simply the result itself.

**Reply by pc223** [reply]
> > But I'm not too sure if this is intended usage, @taooceros can you confirm if this is the "intended" way to use "ContextData"? I don't see this in the docs 😅
> 
> To store information for the specific result. Because context menu is for one specific results, so to display the context menu of the result, we need some information from the result.
> 
> In some Flow built in plugin, the context data is simply the result itself.

So we can add "ContextData" use case like this to the docs and sample plugins right? I just want to make sure this is the right way to use "ContextData" for Python plugins, like it fits with Flow design.

**Reply by taooceros** [reply]
Yeah sure, that will be great. There is no some _right ways_ to use `ContextData`. It is just the place to store information for the context menu in a specific result.

**Reply by taooceros** [reply]
The example of Context Data you display is a list, but in fact, it can be anything. It will be parsed by Json Library so anything (dictionary, object, etc.) can be used in Context Data.

**Reply by pc223** [reply]
> The example of Context Data you display is a list, but in fact, it can be anything. It will be parsed by Json Library so anything (dictionary, object, etc.) can be used in Context Data.

Got it!

**Comment by taooceros** [comment]
By the way, the bug should be fixed? And it won't affect context menu if you do have one.
